File: python/src/wdl_model/model/types/wdl_type.py
"""Base type nodes for the Python WDL object model."""

from __future__ import annotations

from enum import Enum

from wdl_model.model.base import WdlNode


class ComponentType(Enum):
    """High-level type families used for traversal and validation dispatch."""

    PRIMITIVE = "PRIMITIVE"
    ARRAY = "ARRAY"
    MAP = "MAP"
    PAIR = "PAIR"
    TYPE_REFERENCE = "TYPE_REFERENCE"


class WdlType(WdlNode):
    """Base class for WDL types.

    The WDL specification defines primitive types, arrays, maps, pairs, custom type references,
    and optionals. Concrete subclasses represent those categories while this base class carries the
    shared optional marker.
    """

    def __init__(self, optional: bool = False):
        self.optional = optional

    def isOptional(self) -> bool:
        """Return whether this type is optional, as in ``T?``."""
        return self.optional

    def setOptional(self, optional: bool) -> None:
        """Set whether this type is optional, as in ``T?``."""
        self.optional = optional

    def componentType(self) -> ComponentType:
        """Return the broad type family for the current node."""
        raise NotImplementedError

File: java/src/main/java/com/myriad/wdl/model/types/WdlType.java
package com.myriad.wdl.model.types;

import com.myriad.wdl.model.base.WdlNode;

/**
 * Base class for WDL types.
 *
 * <p>The WDL specification defines primitive types, arrays, maps, pairs, custom type references,
 * optionals, and related coercion rules. Concrete subclasses represent those categories while this
 * base class carries the shared optional marker used by the validators and processors.
 */
public abstract class WdlType implements WdlNode {
  /** High-level type family used for traversal and validation dispatch. */
  public static enum ComponentType {
    PRIMITIVE,
    TYPEREF,
    ARRAY,
    PAIR,
    MAP,
  }

  protected boolean optional;

  protected WdlType() {}

  protected WdlType(boolean optional) {
    this.optional = optional;
  }

  /** Returns the broad category of this type node. */
  public abstract ComponentType componentType();

  /** Returns whether this type is optional, as in {@code T?}. */
  public boolean isOptional() {
    return optional;
  }

  /** Sets whether this type is optional, as in {@code T?}. */
  public void setOptional(boolean optional) {
    this.optional = optional;
  }

  @Override
  public String toString() {
    return getClass().getSimpleName();
  }
}
